docs/running-on-kubernetes.md

```diff
@@ -0,0 +1,512 @@
+    Kubernetes master is running at http://127.0.0.1:6443
+
+In the above example, the specific Kubernetes cluster can be used with spark submit by specifying
+`--master k8s://http://127.0.0.1:6443` as an argument to spark-submit. Additionally, it is also possible to use the
+authenticating proxy, `kubectl proxy` to communicate to the Kubernetes API.
+
+The local proxy can be started by:
+
+     kubectl proxy
+
+If the local proxy is running at localhost:8001, `--master k8s://http://127.0.0.1:8001` can be used as the argument to
+spark-submit. Finally, notice that in the above example we specify a jar with a specific URI with a scheme of `local://`.
+This URI is the location of the example jar that is already in the Docker image.
+
+## Dependency Management
+
+If your application's dependencies are all hosted in remote locations like HDFS or http servers, they may be referred to
+by their appropriate remote URIs. Also, application dependencies can be pre-mounted into custom-built Docker images.
+Those dependencies can be added to the classpath by referencing them with `local://` URIs and/or setting the
+`SPARK_EXTRA_CLASSPATH` environment variable in your Dockerfiles.
+
+## Introspection and Debugging
+
+These are the different ways in which you can investigate a running/completed Spark application, monitor progress, and
+take actions.
+
+### Accessing Logs
+
+Logs can be accessed using the kubernetes API and the `kubectl` CLI. When a Spark application is running, it's possible
+to stream logs from the application using:
+
+    kubectl -n=<namespace> logs -f <driver-pod-name>
+
+The same logs can also be accessed through the
+[kubernetes dashboard](https://kubernetes.io/docs/tasks/access-application-cluster/web-ui-dashboard/) if installed on
+the cluster.
+
+### Accessing Driver UI
+
+The UI associated with any application can be accessed locally using
+[`kubectl port-forward`](https://kubernetes.io/docs/tasks/access-application-cluster/port-forward-access-application-cluster/#forward-a-local-port-to-a-port-on-the-pod).
+
+    kubectl port-forward <driver-pod-name> 4040:4040
+
+Then, the spark driver UI can be accessed on `http://localhost:4040`.
+
+### Debugging Failures
+
+There may be several kinds of failures. If the Kubernetes API server rejects the request made from spark-submit, or the
+connection is refused for a different reason, the submission logic should indicate the error encountered. However, if there
+are errors during the running of the application, often, the best way to investigate may be through the kubernetes CLI.
+
+To get some basic information about the scheduling decisions made around the driver pod, you can run:
+
+    kubectl describe pod <spark-driver-pod>
+
+If the pod has encountered a runtime error, the status can be probed further using:
+
+    kubectl logs <spark-driver-pod>
+
+Finally, deleting the driver pod will clean up the entire spark application, includling all executors, associated
```

Comment:
I would add `status and logs of failed executor pods can be checked in similar ways.` at the beginning of this paragraph.

Reply:
Done